Guide/controller.markdown

```diff
@@ -0,0 +1,355 @@
+# Controller & Actions
+
+```toc
+```
+
+## Introduction
+
+In IHP an action is a place for request handling logic. A controller is just a group of related actions.
+
+You can think about actions a message sent to your application, e.g. a `ShowPostAction { postId :: Id Post }` is basically the message "Show me the post with id $postId".
+
+Each action needs to be define as a data structure inside `Web/Types.hs`. Therefore you can see an overview of all the messages which can be sent to your application just by looking at `Web/Types.hs`.
+
+## Creating a new Controller
+
+We recommend to use the code generators for adding a new controller. Using the GUI, you can open [http://localhost:8001/NewController](http://localhost:8001/NewController). Using the CLI run `new-controller CONTROLLER_NAME`.
+
+The following section will guide you through the manual process of creating a new controller.
+
+Let's say we want to create a new controller with a single action `ShowPostAction`. First we need to define our types in `Web/Types.hs`:
+
+```haskell
+data PostsController
+    = ShowPostAction { postId :: !(Id Post) }
+    deriving (Eq, Show, Data)
+```
+
+This define a type `PostsController` with a data constructor `ShowPostAction { postId :: !(Id Post) }`. The arguent `postId` will later be filled with the `postId` parameter of the request url. This is done automatically by the IHP router. IHP also requires the controller to have `Eq`, `Show` and `Data` instances. Therefore we derive them here.
+
+After we have defined the "interface" for our controller, we need to implement the actual request handling logic. IHP expects to find this inside the `action` function of the [`Controller`](https://ihp.digitallyinduced.com/api-docs/IHP-ControllerSupport.html#t:Controller) instance. We can define this instance in `Web/Controller/Posts.hs`:
+
+```haskell
+module Web.Controller.Posts where
+
+import Web.Controller.Prelude
+
+instance Controller PostsController where
+    action ShowPostAction { postId } = renderPlain "Hello World"
+```
+
+This implementation for `ShowPostAction` responds with a simple plain text message. The `action` implementation is usally a big pattern match over all possible actions of a controller.
+
+## Reading Query and Body Parameters
+
+Inside the action you can access request parameters using the `param` function. A parameter can either be a url parameter like `?paramName=paramValue` (*this is also called a query parameter*), or given as a form field like `<form><input name="paramName" value="paramValue"/></form>` (*in that case we're talking about a body parameter*). The `param` function will work with query and body parameters, so you don't have to worry about that.
```

Comment:
What does this do when there are a body and a query parameter with the same name? Maybe not common but good to know.

Reply:
Added
```
in case a query and body parameter is set with the same name, the body parameter will take priority
```
to the markdown :)